{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vignette for projecting data in the Stemformatics Integrated Atlas  \n",
    "Written by Jarny Choi. Last updated 25 May 2021\n",
    "\n",
    "The [Integrated Atlas](http://stemformatics.org/atlas) page in [Stemformatics](http://stemformatics.org) provides a tool to project your own data, enabling a comparison of cell types in the data to those in the atlas. This vignette outlines how this can be done in detail with examples.\n",
    "\n",
    "### Citation\n",
    "If you have found projecting your own data onto the Stemformatics atlas useful, please consider citing:  \n",
    "**A simple, scalable approach to building a cross-platform transcriptome atlas**, Paul W Angel, Nadia Rajab, Yidi Deng, Chris M Pacheco, Tyrone Chen, Kim-Anh Lê Cao, Jarny Choi, Christine A Wells, *PLOS Comp Bio (2020)*, doi: https://doi.org/10.1371/journal.pcbi.1008219\n",
    "\n",
    "\n",
    "## Types of data and format\n",
    "Since the atlas is currently built from microarray and bulk RNA-seq data, these are the primary data types the projection is currently designed for. The projection tool however does not discriminate on the source of data, as long as the row ids are Ensembl ids. Single cell RNA-seq data can be projected, but we have found that aggregating the data to simulate pseudo bulk data yields the best results (see example below). For microarray data, probe ids should be converted to gene ids before projecting.\n",
    "\n",
    "The projection tool also requires a sample table, whose row ids match the columns of the expression matrix. Only one column from this table is required and will be selected for grouping after projection. The user can specify this column - otherwise the first column in the sample table will be used by default.\n",
    "\n",
    "In summary, the projection tool requires 2 text files in tab separated format:  \n",
    "- An expression matrix with human Ensembl ids as row ids and sample ids as column names.\n",
    "- A sample matrix with sample ids as row ids, matching the columns of the expression matrix (not necessarily in the same order). Only one column is required in this matrix.\n",
    "\n",
    "During the projection, the input expression matrix will be filtered on genes which are present in the atlas, then rank transformed before projection is performed. If less than 50% of the genes in the atlas are present in the input matrix, the projection will not proceed and an error will result. The genes used by the atlas can be downloaded as a tab separated file from the download data function on the atlas page.\n",
    "\n",
    "## Example projection (bulk RNA-Seq)\n",
    "In this example, we will project blood expression data onto the Stemformatics [Blood Atlas](http://stemformatics.org/atlas/blood) and see if the projected samples are located close to the matching counterparts in the atlas on the PCA plot. Our input data come from [haemosphere.org](http://haemosphere.org), which contains a selection of gene expression datasets specific to blood. In particular, we focus on **Haemopedia-Human-RNASeq** dataset, which contains samples from various blood lineages, including lymphocytes and myeloid cells.\n",
    "\n",
    "The two files we need can be downloaded by clicking on the “download” button from haemosphere’s [datasets page](http://haemosphere.org/datasets/show) for **Haemopedia-Human-RNASeq** dataset. We can download “cpm values” and “sample table” files. Note that since these expression values will be rank transformed per sample independently of all other samples during the projection, the exact choice of normalisation used as input file is not so important.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>myDC123.2</th>\n",
       "      <th>myDC123.1</th>\n",
       "      <th>MonoNonClassic.1</th>\n",
       "      <th>CD4T.1</th>\n",
       "      <th>CD8T.1</th>\n",
       "      <th>Eo.1</th>\n",
       "      <th>myDC.1</th>\n",
       "      <th>MemB.1</th>\n",
       "      <th>Mono.1</th>\n",
       "      <th>NveB.1</th>\n",
       "      <th>...</th>\n",
       "      <th>NveB.4</th>\n",
       "      <th>NK.4</th>\n",
       "      <th>CD4T.5</th>\n",
       "      <th>CD8T.5</th>\n",
       "      <th>myDC.3</th>\n",
       "      <th>Mono.5</th>\n",
       "      <th>NveB.5</th>\n",
       "      <th>Neut.3</th>\n",
       "      <th>NK.5</th>\n",
       "      <th>pDC.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genenames</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1.224177</td>\n",
       "      <td>1.110908</td>\n",
       "      <td>4.317801</td>\n",
       "      <td>5.906965</td>\n",
       "      <td>2.579694</td>\n",
       "      <td>2.76102</td>\n",
       "      <td>1.205595</td>\n",
       "      <td>1.890113</td>\n",
       "      <td>3.837349</td>\n",
       "      <td>2.994901</td>\n",
       "      <td>...</td>\n",
       "      <td>2.893165</td>\n",
       "      <td>4.161794</td>\n",
       "      <td>5.296969</td>\n",
       "      <td>5.382940</td>\n",
       "      <td>5.235310</td>\n",
       "      <td>3.499009</td>\n",
       "      <td>4.533116</td>\n",
       "      <td>1.164611</td>\n",
       "      <td>4.606214</td>\n",
       "      <td>3.565716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>0.087441</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.160731</td>\n",
       "      <td>0.189172</td>\n",
       "      <td>0.068792</td>\n",
       "      <td>0.139817</td>\n",
       "      <td>0.090264</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.075866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000274890</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 myDC123.2  myDC123.1  MonoNonClassic.1    CD4T.1    CD8T.1  \\\n",
       "genenames                                                                     \n",
       "ENSG00000223972   0.000000   0.000000          0.000000  0.000000  0.000000   \n",
       "ENSG00000227232   1.224177   1.110908          4.317801  5.906965  2.579694   \n",
       "ENSG00000278267   0.087441   0.000000          0.000000  0.000000  0.000000   \n",
       "ENSG00000243485   0.000000   0.000000          0.000000  0.000000  0.000000   \n",
       "ENSG00000274890   0.000000   0.000000          0.000000  0.000000  0.000000   \n",
       "\n",
       "                    Eo.1    myDC.1    MemB.1    Mono.1    NveB.1  ...  \\\n",
       "genenames                                                         ...   \n",
       "ENSG00000223972  0.00000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSG00000227232  2.76102  1.205595  1.890113  3.837349  2.994901  ...   \n",
       "ENSG00000278267  0.00000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSG00000243485  0.00000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSG00000274890  0.00000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "\n",
       "                   NveB.4      NK.4    CD4T.5    CD8T.5    myDC.3    Mono.5  \\\n",
       "genenames                                                                     \n",
       "ENSG00000223972  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSG00000227232  2.893165  4.161794  5.296969  5.382940  5.235310  3.499009   \n",
       "ENSG00000278267  0.160731  0.189172  0.068792  0.139817  0.090264  0.000000   \n",
       "ENSG00000243485  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSG00000274890  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                   NveB.5    Neut.3      NK.5     pDC.2  \n",
       "genenames                                                \n",
       "ENSG00000223972  0.000000  0.000000  0.000000  0.000000  \n",
       "ENSG00000227232  4.533116  1.164611  4.606214  3.565716  \n",
       "ENSG00000278267  0.000000  0.000000  0.000000  0.075866  \n",
       "ENSG00000243485  0.000000  0.000000  0.000000  0.000000  \n",
       "ENSG00000274890  0.000000  0.000000  0.000000  0.000000  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# The downloaded expression matrix looks like this:\n",
    "import pandas\n",
    "\n",
    "hp = pandas.read_csv(\"Haemopedia-Human-RNASeq_cpm.txt\", sep=\"\\t\", index_col=0)\n",
    "display(hp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellTypeAbbreviation</th>\n",
       "      <th>celltype</th>\n",
       "      <th>organism</th>\n",
       "      <th>long celltype</th>\n",
       "      <th>cell_lineage</th>\n",
       "      <th>immunophenotype</th>\n",
       "      <th>tissue</th>\n",
       "      <th>donor</th>\n",
       "      <th>totalReads</th>\n",
       "      <th>mappedPercent</th>\n",
       "      <th>assignedPercent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MemB.1</th>\n",
       "      <td>MemB</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD3:- CD19:+ CD27:+</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Donor1</td>\n",
       "      <td>18699477</td>\n",
       "      <td>89.24</td>\n",
       "      <td>63.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MemB.2</th>\n",
       "      <td>MemB</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD3:- CD19:+ CD27:+</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Donor2</td>\n",
       "      <td>17747099</td>\n",
       "      <td>89.89</td>\n",
       "      <td>60.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MemB.3</th>\n",
       "      <td>MemB</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Memory B Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD3:- CD19:+ CD27:+</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Donor3</td>\n",
       "      <td>17434867</td>\n",
       "      <td>90.06</td>\n",
       "      <td>63.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NveB.1</th>\n",
       "      <td>NveB</td>\n",
       "      <td>Naive B Cell</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Naive B Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD3:- CD19:+ CD27:-</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Donor1</td>\n",
       "      <td>18521249</td>\n",
       "      <td>90.24</td>\n",
       "      <td>59.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NveB.2</th>\n",
       "      <td>NveB</td>\n",
       "      <td>Naive B Cell</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Naive B Cell</td>\n",
       "      <td>B Cell Lineage</td>\n",
       "      <td>CD3:- CD19:+ CD27:-</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Donor2</td>\n",
       "      <td>17695868</td>\n",
       "      <td>90.46</td>\n",
       "      <td>62.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         cellTypeAbbreviation       celltype      organism  long celltype  \\\n",
       "sampleId                                                                    \n",
       "MemB.1                   MemB  Memory B Cell  Homo Sapiens  Memory B Cell   \n",
       "MemB.2                   MemB  Memory B Cell  Homo Sapiens  Memory B Cell   \n",
       "MemB.3                   MemB  Memory B Cell  Homo Sapiens  Memory B Cell   \n",
       "NveB.1                   NveB   Naive B Cell  Homo Sapiens   Naive B Cell   \n",
       "NveB.2                   NveB   Naive B Cell  Homo Sapiens   Naive B Cell   \n",
       "\n",
       "            cell_lineage      immunophenotype            tissue   donor  \\\n",
       "sampleId                                                                  \n",
       "MemB.1    B Cell Lineage  CD3:- CD19:+ CD27:+  Peripheral Blood  Donor1   \n",
       "MemB.2    B Cell Lineage  CD3:- CD19:+ CD27:+  Peripheral Blood  Donor2   \n",
       "MemB.3    B Cell Lineage  CD3:- CD19:+ CD27:+  Peripheral Blood  Donor3   \n",
       "NveB.1    B Cell Lineage  CD3:- CD19:+ CD27:-  Peripheral Blood  Donor1   \n",
       "NveB.2    B Cell Lineage  CD3:- CD19:+ CD27:-  Peripheral Blood  Donor2   \n",
       "\n",
       "          totalReads  mappedPercent  assignedPercent  \n",
       "sampleId                                              \n",
       "MemB.1      18699477          89.24            63.72  \n",
       "MemB.2      17747099          89.89            60.29  \n",
       "MemB.3      17434867          90.06            63.66  \n",
       "NveB.1      18521249          90.24            59.43  \n",
       "NveB.2      17695868          90.46            62.28  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# And the sample matrix looks like this\n",
    "hpSamples = pandas.read_csv(\"Haemopedia-Human-RNASeq_samples.txt\", sep=\"\\t\", index_col=0)\n",
    "display(hpSamples.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that even though row ids of the sample table should contain all column names of the expression matrix, they do not need to be in the same order.\n",
    "\n",
    "Now we perform the projection using “other functions” >> “project other data” >> “Your own dataset” function on the atlas page, where it asks to provide these two files, together with a name which will be used as a prefix to the projected samples for easier identification (we use “hp” as the name in this example). We also leave the sample column field blank here, which indicates we will use the first column of the sample table to group the projected samples together. \n",
    "\n",
    "![Atlas projection dialog](Stemformatics_atlas_projection_dialog.png)\n",
    "\n",
    "The projection process may take some time - it will scale with the expression file size being uploaded. In this example, the resulting projection shows a good biological concordance between the projected samples and cell types in the atlas, giving us some confidence that projection has worked well.\n",
    "\n",
    "![Haemopedia projection](Stemformatics_blood_atlas_projection_Haemopedia.jpeg)\n",
    "\n",
    "*The samples from Haemopedia-Human-RNASeq data (diamond shapes) project onto the Stemformatics Blood Atlas near expected cell types. It is easier to see this by using the atlas page directly for projection using the files downloaded.*\n",
    "\n",
    "## Projecting single cell data\n",
    "\n",
    "Single cell RNA-seq data can be projected onto the atlas, but we have found that aggregating the data beforehand yields much better results than naively projecting each sample onto the atlas. This is due to the difference in the data structure between bulk and single cell RNA-seq data. There will also be performance issues for very large single cell datasets, both in terms of file upload and for processing if aggregation is not done.\n",
    "\n",
    "As an example, we take this study by [Villani, et. al.](https://science.sciencemag.org/content/356/6335/eaah4573) where they profiled human dendritic cells and monocytes to find sub clusters. We can download the files easily through the Broad Institute’s [Single Cell Data Portal](https://singlecell.broadinstitute.org/single_cell/study/SCP43/atlas-of-human-blood-dendritic-cells-and-monocytes#study-download) (note - you need to register and sign-in before downloading). We download “expression_matrix_tpm.txt” and “metadata.txt” files.\n",
    "\n",
    "Note that the expression matrix here has gene symbols, rather than Ensembl gene ids, so we need to convert these first before projecting onto the atlas. There are various ways of obtaining the mapping between Ensembl gene ids to gene symbols, including using biomart. One simple option is to simply use the file provided on the atlas page :\"other functions\" >> \"download data/plots\" >> \"genes matrix\", which contains this mapping.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD141_P10_S73</th>\n",
       "      <th>CD141_P10_S74</th>\n",
       "      <th>CD141_P10_S75</th>\n",
       "      <th>CD141_P10_S76</th>\n",
       "      <th>CD141_P10_S77</th>\n",
       "      <th>CD141_P10_S78</th>\n",
       "      <th>CD141_P10_S79</th>\n",
       "      <th>CD141_P10_S80</th>\n",
       "      <th>CD141_P10_S81</th>\n",
       "      <th>CD141_P10_S82</th>\n",
       "      <th>...</th>\n",
       "      <th>Mono_nonclassical_S91</th>\n",
       "      <th>Mono_nonclassical_S92</th>\n",
       "      <th>Mono_nonclassical_S93</th>\n",
       "      <th>Mono_nonclassical_S94</th>\n",
       "      <th>Mono_nonclassical_S95</th>\n",
       "      <th>Mono_nonclassical_S96</th>\n",
       "      <th>Mono_nonclassical_S97</th>\n",
       "      <th>Mono_nonclassical_S98</th>\n",
       "      <th>Mono_nonclassical_S99</th>\n",
       "      <th>Mono_nonclassical_S9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1/2-SBSRNA4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.208960</td>\n",
       "      <td>1.853168</td>\n",
       "      <td>1.261298</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.667829</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.534714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.936093</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5S_RRNA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.029650</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_8S_RRNA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.152456</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>4.625365</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.549975</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.035434</td>\n",
       "      <td>2.436241</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.978077</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.355502</td>\n",
       "      <td>3.64545</td>\n",
       "      <td>3.844814</td>\n",
       "      <td>3.78646</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.670694</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1078 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CD141_P10_S73  CD141_P10_S74  CD141_P10_S75  CD141_P10_S76  \\\n",
       "GENE                                                                      \n",
       "1/2-SBSRNA4       0.000000            0.0            0.0            0.0   \n",
       "5S_RRNA           0.000000            0.0            0.0            0.0   \n",
       "5_8S_RRNA         0.000000            0.0            0.0            0.0   \n",
       "7SK               4.625365            0.0            0.0            0.0   \n",
       "A1BG              2.670694            0.0            0.0            0.0   \n",
       "\n",
       "             CD141_P10_S77  CD141_P10_S78  CD141_P10_S79  CD141_P10_S80  \\\n",
       "GENE                                                                      \n",
       "1/2-SBSRNA4       1.208960       1.853168       1.261298       0.000000   \n",
       "5S_RRNA           0.000000       0.000000       0.000000       3.029650   \n",
       "5_8S_RRNA         0.000000       0.000000       4.152456       0.000000   \n",
       "7SK               4.549975       0.000000       3.035434       2.436241   \n",
       "A1BG              0.000000       0.000000       0.000000       0.000000   \n",
       "\n",
       "             CD141_P10_S81  CD141_P10_S82  ...  Mono_nonclassical_S91  \\\n",
       "GENE                                       ...                          \n",
       "1/2-SBSRNA4       0.667829            0.0  ...               1.534714   \n",
       "5S_RRNA           0.000000            0.0  ...               0.000000   \n",
       "5_8S_RRNA         0.000000            0.0  ...               0.000000   \n",
       "7SK               0.000000            0.0  ...               0.000000   \n",
       "A1BG              0.000000            0.0  ...               0.000000   \n",
       "\n",
       "             Mono_nonclassical_S92  Mono_nonclassical_S93  \\\n",
       "GENE                                                        \n",
       "1/2-SBSRNA4                    0.0               0.000000   \n",
       "5S_RRNA                        0.0               0.000000   \n",
       "5_8S_RRNA                      0.0               0.000000   \n",
       "7SK                            0.0               2.978077   \n",
       "A1BG                           0.0               0.000000   \n",
       "\n",
       "             Mono_nonclassical_S94  Mono_nonclassical_S95  \\\n",
       "GENE                                                        \n",
       "1/2-SBSRNA4                    0.0                    0.0   \n",
       "5S_RRNA                        0.0                    0.0   \n",
       "5_8S_RRNA                      0.0                    0.0   \n",
       "7SK                            0.0                    0.0   \n",
       "A1BG                           0.0                    0.0   \n",
       "\n",
       "             Mono_nonclassical_S96  Mono_nonclassical_S97  \\\n",
       "GENE                                                        \n",
       "1/2-SBSRNA4               0.936093                0.00000   \n",
       "5S_RRNA                   0.000000                0.00000   \n",
       "5_8S_RRNA                 0.000000                0.00000   \n",
       "7SK                       3.355502                3.64545   \n",
       "A1BG                      0.000000                0.00000   \n",
       "\n",
       "             Mono_nonclassical_S98  Mono_nonclassical_S99  \\\n",
       "GENE                                                        \n",
       "1/2-SBSRNA4               0.000000                0.00000   \n",
       "5S_RRNA                   0.000000                0.00000   \n",
       "5_8S_RRNA                 0.000000                0.00000   \n",
       "7SK                       3.844814                3.78646   \n",
       "A1BG                      0.000000                0.00000   \n",
       "\n",
       "             Mono_nonclassical_S9  \n",
       "GENE                               \n",
       "1/2-SBSRNA4                   0.0  \n",
       "5S_RRNA                       0.0  \n",
       "5_8S_RRNA                     0.0  \n",
       "7SK                           0.0  \n",
       "A1BG                          0.0  \n",
       "\n",
       "[5 rows x 1078 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TYPE</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD141_P10_S73</th>\n",
       "      <td>DC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD141_P10_S74</th>\n",
       "      <td>DC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD141_P10_S75</th>\n",
       "      <td>DC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD141_P10_S76</th>\n",
       "      <td>DC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD141_P10_S77</th>\n",
       "      <td>DC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              group\n",
       "TYPE               \n",
       "CD141_P10_S73   DC1\n",
       "CD141_P10_S74   DC5\n",
       "CD141_P10_S75   DC1\n",
       "CD141_P10_S76   DC1\n",
       "CD141_P10_S77   DC1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All columns of expression matrix are in samples: True\n"
     ]
    }
   ],
   "source": [
    "# Read the downloaded expression matrix and samples files. Samples file has two rows of headers so we can skip one.\n",
    "villani = pandas.read_csv(\"expression_matrix_tpm.txt\", sep=\"\\t\", index_col=0)\n",
    "villaniSamples = pandas.read_csv(\"metadata.txt\", sep=\"\\t\", index_col=0, skiprows=1)\n",
    "\n",
    "display(villani.head())\n",
    "display(villaniSamples.head())\n",
    "\n",
    "print(\"All columns of expression matrix are in samples:\", set(villani.columns).issubset(set(villaniSamples.index)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Even though we will aggregate the samples before projecting, we can see what happens first without any aggregation as a comparison. So first convert the gene symbols to Ensembl ids and save this file. Also remove one extra row of headers from sample table as that won't work for projection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Function to convert row ids in pandas DataFrame df from gene symbols to Ensembl gene ids.\n",
    "Uses the gene annotation file downloaded from Stemformatics atlas page.\n",
    "\"\"\"\n",
    "def convertGeneSymbolsToEnsemblIds(df, geneAnnotationFile=\"blood_atlas_genes_v7.1.tsv\"):\n",
    "    genes = pandas.read_csv(geneAnnotationFile, sep=\"\\t\").set_index(\"symbol\")\n",
    "    commonSymbols = list(set(genes.index).intersection(set(df.index)))\n",
    "\n",
    "    # drop duplicate index from expression before running reindex\n",
    "    df = df.loc[~df.index.duplicated(keep='first')].reindex(commonSymbols)\n",
    "    df.index = genes.loc[commonSymbols][\"ensembl\"]\n",
    "    return df\n",
    "\n",
    "# Use the function to save Ensembl id version of villani data. Also save sample table with fixed headers.\n",
    "convertGeneSymbolsToEnsemblIds(villani).to_csv(\"expression_matrix_tpm_EnsemblIds.txt\", sep=\"\\t\")\n",
    "villaniSamples.to_csv(\"metadata_fixed_header.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we project this unaggregated data directly onto the atlas, the projected samples tend to cluster together rather than show sufficient similarity with atlas cell types to make inferences.\n",
    "\n",
    "![Villani projection](Stemformatics_blood_atlas_projection_villani_original.jpeg)\n",
    "\n",
    "\n",
    "Now we aggregate based on the clusters which are already assigned by the second column of metadata.txt file. The number of individual samples to aggregate within a cluster can be determined in various ways, but we have found that aggregating between 5-8 cells is sufficient to approximate the distribution of bulk data. More details on this can be found in [this paper](https://www.biorxiv.org/content/10.1101/719237v2) (See supplementary methods where it talks about using Kolmogorov–Smirnov statistic to work out the optimal aggregation size using this same dataset).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26593, 185)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DC1_0</th>\n",
       "      <th>DC1_1</th>\n",
       "      <th>DC1_2</th>\n",
       "      <th>DC1_3</th>\n",
       "      <th>DC1_4</th>\n",
       "      <th>DC1_5</th>\n",
       "      <th>DC1_6</th>\n",
       "      <th>DC1_7</th>\n",
       "      <th>DC1_8</th>\n",
       "      <th>DC1_9</th>\n",
       "      <th>...</th>\n",
       "      <th>DC6_25</th>\n",
       "      <th>DC6_26</th>\n",
       "      <th>DC6_27</th>\n",
       "      <th>DC6_28</th>\n",
       "      <th>Mono3_0</th>\n",
       "      <th>Mono3_1</th>\n",
       "      <th>Mono3_2</th>\n",
       "      <th>Mono3_3</th>\n",
       "      <th>Mono3_4</th>\n",
       "      <th>Mono3_5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1/2-SBSRNA4</th>\n",
       "      <td>5.651161</td>\n",
       "      <td>2.032088</td>\n",
       "      <td>1.885553</td>\n",
       "      <td>8.382685</td>\n",
       "      <td>3.210061</td>\n",
       "      <td>3.040054</td>\n",
       "      <td>5.545655</td>\n",
       "      <td>7.718105</td>\n",
       "      <td>8.314636</td>\n",
       "      <td>5.579928</td>\n",
       "      <td>...</td>\n",
       "      <td>5.998881</td>\n",
       "      <td>4.023385</td>\n",
       "      <td>0.916291</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.949551</td>\n",
       "      <td>7.380411</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.249040</td>\n",
       "      <td>3.85469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5S_RRNA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.044279</td>\n",
       "      <td>4.477564</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.291010</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.044279</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.057964</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_8S_RRNA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.183728</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>11.921593</td>\n",
       "      <td>3.379633</td>\n",
       "      <td>4.850858</td>\n",
       "      <td>16.950666</td>\n",
       "      <td>4.942071</td>\n",
       "      <td>7.402615</td>\n",
       "      <td>9.517593</td>\n",
       "      <td>8.238556</td>\n",
       "      <td>2.282382</td>\n",
       "      <td>8.538484</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.517498</td>\n",
       "      <td>7.053788</td>\n",
       "      <td>4.795046</td>\n",
       "      <td>12.975854</td>\n",
       "      <td>4.965568</td>\n",
       "      <td>7.102106</td>\n",
       "      <td>3.789177</td>\n",
       "      <td>14.023648</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>1.633154</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.091674</td>\n",
       "      <td>2.070653</td>\n",
       "      <td>0.285179</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.070653</td>\n",
       "      <td>5.439958</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.401218</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.787763</td>\n",
       "      <td>3.281663</td>\n",
       "      <td>3.281663</td>\n",
       "      <td>4.783023</td>\n",
       "      <td>3.281663</td>\n",
       "      <td>4.621107</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 185 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 DC1_0     DC1_1     DC1_2      DC1_3     DC1_4     DC1_5  \\\n",
       "GENE                                                                        \n",
       "1/2-SBSRNA4   5.651161  2.032088  1.885553   8.382685  3.210061  3.040054   \n",
       "5S_RRNA       0.000000  4.044279  4.477564   0.000000  0.000000  3.291010   \n",
       "5_8S_RRNA     0.000000  0.000000  0.000000   4.183728  0.000000  0.000000   \n",
       "7SK          11.921593  3.379633  4.850858  16.950666  4.942071  7.402615   \n",
       "A1BG          1.633154  0.000000  4.091674   2.070653  0.285179  0.000000   \n",
       "\n",
       "                DC1_6     DC1_7     DC1_8     DC1_9  ...    DC6_25    DC6_26  \\\n",
       "GENE                                                 ...                       \n",
       "1/2-SBSRNA4  5.545655  7.718105  8.314636  5.579928  ...  5.998881  4.023385   \n",
       "5S_RRNA      0.000000  0.000000  0.000000  4.044279  ...  0.000000  0.000000   \n",
       "5_8S_RRNA    0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "7SK          9.517593  8.238556  2.282382  8.538484  ...  0.000000  3.517498   \n",
       "A1BG         0.000000  2.070653  5.439958  0.000000  ...  0.000000  7.401218   \n",
       "\n",
       "               DC6_27    DC6_28    Mono3_0   Mono3_1   Mono3_2   Mono3_3  \\\n",
       "GENE                                                                       \n",
       "1/2-SBSRNA4  0.916291  0.000000   4.949551  7.380411  0.000000  0.000000   \n",
       "5S_RRNA      0.000000  0.000000   0.000000  0.000000  0.000000  5.057964   \n",
       "5_8S_RRNA    0.000000  0.000000   0.000000  0.000000  0.000000  0.000000   \n",
       "7SK          7.053788  4.795046  12.975854  4.965568  7.102106  3.789177   \n",
       "A1BG         0.000000  3.787763   3.281663  3.281663  4.783023  3.281663   \n",
       "\n",
       "               Mono3_4  Mono3_5  \n",
       "GENE                             \n",
       "1/2-SBSRNA4   3.249040  3.85469  \n",
       "5S_RRNA       0.000000  0.00000  \n",
       "5_8S_RRNA     0.000000  0.00000  \n",
       "7SK          14.023648  0.00000  \n",
       "A1BG          4.621107  0.00000  \n",
       "\n",
       "[5 rows x 185 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Function to gggregate pandas DataFrame df and return the aggregated data frame and a matching sample table.\n",
    "    Example: dfNew, samplesNew = aggregateExpressionMatrix(df, [\"HSC\",\"Mono\",\"HSC\",...])\n",
    "groups is a list of sample group items, same length as columns of df, denoting which sample belongs to which group.\n",
    "size is the number of samples in each group to sum over - random selection with replacement will be used.\n",
    "\"\"\"\n",
    "import numpy\n",
    "\n",
    "def aggregateExpressionMatrix(df, groups, size=6):\n",
    "    \n",
    "    # Convert the list into pandas Series to take advantage of its features\n",
    "    groups = pandas.Series(groups)\n",
    "    \n",
    "    # Aggregated data frame and new sample group membership to return\n",
    "    aggregated = pandas.DataFrame()\n",
    "    aggSampleGroup = []\n",
    "    \n",
    "    for group in groups.unique():  # loop through each unique item in groups\n",
    "        \n",
    "        # Indices of groups which match this group\n",
    "        matchingIndex = groups[groups==group].index\n",
    "        \n",
    "        # How many aggregated samples should we produce per group? \n",
    "        # Here we divide number of samples in the group by size. Take size if this number is smaller.\n",
    "        numberOfSamples = max(round(len(matchingIndex)/size), size)\n",
    "                \n",
    "        for i in range(numberOfSamples):\n",
    "            # Select a subset of matchinIndex randomly with replacement. So chosenIndex may have duplicates.\n",
    "            chosenIndex = numpy.random.choice(matchingIndex, size)\n",
    "\n",
    "            # new sample id\n",
    "            sampleId = \"%s_%s\" % (group, i)\n",
    "            aggSampleGroup.append([sampleId, group])\n",
    "            \n",
    "            # Construct a column \"subset\" of df using chosenIndex, where duplicates may be present,\n",
    "            # then sum these column-wise and add this to aggregated\n",
    "            subset = df.iloc[:,chosenIndex]\n",
    "            aggregated[sampleId] = subset.sum(axis=1)\n",
    "        \n",
    "    return aggregated, pandas.DataFrame(aggSampleGroup, columns=['sampleId', 'group']).set_index('sampleId')\n",
    "    \n",
    "villaniAggregated, villaniAggSamples = aggregateExpressionMatrix(villani, villaniSamples[\"group\"].tolist())\n",
    "print(villaniAggregated.shape)\n",
    "display(villaniAggregated.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DC1_0</th>\n",
       "      <th>DC1_1</th>\n",
       "      <th>DC1_2</th>\n",
       "      <th>DC1_3</th>\n",
       "      <th>DC1_4</th>\n",
       "      <th>DC1_5</th>\n",
       "      <th>DC1_6</th>\n",
       "      <th>DC1_7</th>\n",
       "      <th>DC1_8</th>\n",
       "      <th>DC1_9</th>\n",
       "      <th>...</th>\n",
       "      <th>DC6_25</th>\n",
       "      <th>DC6_26</th>\n",
       "      <th>DC6_27</th>\n",
       "      <th>DC6_28</th>\n",
       "      <th>Mono3_0</th>\n",
       "      <th>Mono3_1</th>\n",
       "      <th>Mono3_2</th>\n",
       "      <th>Mono3_3</th>\n",
       "      <th>Mono3_4</th>\n",
       "      <th>Mono3_5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000116698</th>\n",
       "      <td>1.047319</td>\n",
       "      <td>8.749755</td>\n",
       "      <td>1.388791</td>\n",
       "      <td>5.990058</td>\n",
       "      <td>3.913622</td>\n",
       "      <td>6.092775</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.319374</td>\n",
       "      <td>1.169381</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.452496</td>\n",
       "      <td>7.715534</td>\n",
       "      <td>8.329410</td>\n",
       "      <td>3.704871</td>\n",
       "      <td>13.112474</td>\n",
       "      <td>14.401569</td>\n",
       "      <td>19.347299</td>\n",
       "      <td>12.298960</td>\n",
       "      <td>7.207455</td>\n",
       "      <td>16.023954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000088205</th>\n",
       "      <td>16.607080</td>\n",
       "      <td>7.829327</td>\n",
       "      <td>12.671372</td>\n",
       "      <td>15.057319</td>\n",
       "      <td>8.433056</td>\n",
       "      <td>13.987898</td>\n",
       "      <td>5.140628</td>\n",
       "      <td>11.858718</td>\n",
       "      <td>9.835893</td>\n",
       "      <td>8.760485</td>\n",
       "      <td>...</td>\n",
       "      <td>24.938408</td>\n",
       "      <td>16.630552</td>\n",
       "      <td>9.860245</td>\n",
       "      <td>5.396706</td>\n",
       "      <td>16.807234</td>\n",
       "      <td>12.230230</td>\n",
       "      <td>1.701105</td>\n",
       "      <td>12.272025</td>\n",
       "      <td>15.599035</td>\n",
       "      <td>3.936638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000077327</th>\n",
       "      <td>2.215948</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.841660</td>\n",
       "      <td>0.536493</td>\n",
       "      <td>1.249988</td>\n",
       "      <td>1.268861</td>\n",
       "      <td>0.978326</td>\n",
       "      <td>4.281652</td>\n",
       "      <td>1.121678</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.303744</td>\n",
       "      <td>3.050629</td>\n",
       "      <td>3.961448</td>\n",
       "      <td>1.371181</td>\n",
       "      <td>3.603027</td>\n",
       "      <td>2.749960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000091039</th>\n",
       "      <td>12.078762</td>\n",
       "      <td>8.438128</td>\n",
       "      <td>11.947039</td>\n",
       "      <td>16.719615</td>\n",
       "      <td>7.717399</td>\n",
       "      <td>10.034621</td>\n",
       "      <td>6.374665</td>\n",
       "      <td>9.214563</td>\n",
       "      <td>6.977688</td>\n",
       "      <td>17.548167</td>\n",
       "      <td>...</td>\n",
       "      <td>21.513627</td>\n",
       "      <td>14.539010</td>\n",
       "      <td>16.512262</td>\n",
       "      <td>9.797026</td>\n",
       "      <td>18.231193</td>\n",
       "      <td>27.684564</td>\n",
       "      <td>21.584457</td>\n",
       "      <td>23.469503</td>\n",
       "      <td>15.274974</td>\n",
       "      <td>21.452959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000118503</th>\n",
       "      <td>5.951538</td>\n",
       "      <td>4.904341</td>\n",
       "      <td>7.379061</td>\n",
       "      <td>6.794721</td>\n",
       "      <td>8.898817</td>\n",
       "      <td>2.656055</td>\n",
       "      <td>3.021186</td>\n",
       "      <td>4.634719</td>\n",
       "      <td>3.621612</td>\n",
       "      <td>7.817818</td>\n",
       "      <td>...</td>\n",
       "      <td>17.674313</td>\n",
       "      <td>12.440998</td>\n",
       "      <td>10.070372</td>\n",
       "      <td>13.044396</td>\n",
       "      <td>19.345186</td>\n",
       "      <td>19.756043</td>\n",
       "      <td>22.068908</td>\n",
       "      <td>15.281602</td>\n",
       "      <td>22.501590</td>\n",
       "      <td>12.960494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 185 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     DC1_0     DC1_1      DC1_2      DC1_3     DC1_4  \\\n",
       "ensembl                                                                \n",
       "ENSG00000116698   1.047319  8.749755   1.388791   5.990058  3.913622   \n",
       "ENSG00000088205  16.607080  7.829327  12.671372  15.057319  8.433056   \n",
       "ENSG00000077327   2.215948  0.000000   0.000000   1.841660  0.536493   \n",
       "ENSG00000091039  12.078762  8.438128  11.947039  16.719615  7.717399   \n",
       "ENSG00000118503   5.951538  4.904341   7.379061   6.794721  8.898817   \n",
       "\n",
       "                     DC1_5     DC1_6      DC1_7     DC1_8      DC1_9  ...  \\\n",
       "ensembl                                                               ...   \n",
       "ENSG00000116698   6.092775  0.000000   5.319374  1.169381   0.000000  ...   \n",
       "ENSG00000088205  13.987898  5.140628  11.858718  9.835893   8.760485  ...   \n",
       "ENSG00000077327   1.249988  1.268861   0.978326  4.281652   1.121678  ...   \n",
       "ENSG00000091039  10.034621  6.374665   9.214563  6.977688  17.548167  ...   \n",
       "ENSG00000118503   2.656055  3.021186   4.634719  3.621612   7.817818  ...   \n",
       "\n",
       "                    DC6_25     DC6_26     DC6_27     DC6_28    Mono3_0  \\\n",
       "ensembl                                                                  \n",
       "ENSG00000116698   7.452496   7.715534   8.329410   3.704871  13.112474   \n",
       "ENSG00000088205  24.938408  16.630552   9.860245   5.396706  16.807234   \n",
       "ENSG00000077327   0.000000   0.000000   0.000000   0.000000   2.303744   \n",
       "ENSG00000091039  21.513627  14.539010  16.512262   9.797026  18.231193   \n",
       "ENSG00000118503  17.674313  12.440998  10.070372  13.044396  19.345186   \n",
       "\n",
       "                   Mono3_1    Mono3_2    Mono3_3    Mono3_4    Mono3_5  \n",
       "ensembl                                                                 \n",
       "ENSG00000116698  14.401569  19.347299  12.298960   7.207455  16.023954  \n",
       "ENSG00000088205  12.230230   1.701105  12.272025  15.599035   3.936638  \n",
       "ENSG00000077327   3.050629   3.961448   1.371181   3.603027   2.749960  \n",
       "ENSG00000091039  27.684564  21.584457  23.469503  15.274974  21.452959  \n",
       "ENSG00000118503  19.756043  22.068908  15.281602  22.501590  12.960494  \n",
       "\n",
       "[5 rows x 185 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convert aggregated matrix to use Ensembl ids. We didn't do this before aggregation since some genes will drop out \n",
    "# in the process of mapping to gene ids, and we want to keep as many genes as we can during the aggregation process.\n",
    "villaniAggregated = convertGeneSymbolsToEnsemblIds(villaniAggregated)\n",
    "display(villaniAggregated.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can now write this aggregated matrix to file, together with new sample table\n",
    "villaniAggregated.to_csv(\"villani_aggregated_expression.tsv\", sep=\"\\t\")\n",
    "villaniAggSamples.to_csv(\"villani_aggregated_samples.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using these files now for projection yields a better result.\n",
    "\n",
    "![Villani projection](Stemformatics_blood_atlas_projection_villani_aggregated.jpeg)\n",
    "\n",
    "*The samples from Villani data after aggregation (diamond shapes) project onto the Stemformatics Blood Atlas. It is easier to explore this on the web page directly after projection. This plot is provided here for a quick reference.*\n",
    "\n",
    "## Interpreting the projection and limitations\n",
    "Projections show the transcriptional similarity between atlas samples and projected samples within the PCA framework. Hence if a projected sample is close to an atlas sample in the PCA plot, it indicates that genes in the atlas with high loadings for the first few principal components share a similar expression pattern in the projected sample after a rank transformation. Note that if a sample is very different to all the samples in the atlas, it will tend to sit in the middle of the plot. This is usually what happens when we project single cell RNA-seq data without any aggregation, due to its very different data structure.\n",
    "\n",
    "Currently there is no tool provided online in Stemformatics to go beyond the projected points to show exact distances or cell type classifications (work in progress). Hence the projections are semi-quantitative since one makes mostly qualitative assessments post-projection, and it is meant to serve as a hypothesis generator, rather than to act as a definitive cell type classifier. Note however that other tools do exist, where the atlas data can serve as a reference for classifier programs. For example, we have used [capybara](https://www.biorxiv.org/content/10.1101/2020.02.17.947390v2) to classify the cell types from the Villani data against the cell types in the atlas - details can be found in [this paper](https://doi.org/10.1016/j.stemcr.2021.04.010).\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
